### PR TITLE
Refactoring MSUtil class

### PR DESCRIPTION
**Pull request by bmourat**
MSUtil class refactored into different categories such as:
MSUtility+Application
MSUtility+Environment
MSUtility+Date
Every category has it own h/m files, so if someone wants to update one specific part of the utility class he can just update the category related to this change.

### CODE REVIEW

File: MobileCenter/MobileCenter/Internals/MobileCenter+Internal.h [hunk 1]
```diff
@@ -9,7 +9,9 @@
 #import "Model/MSLogWithProperties.h"
 #import "Model/Utils/MSUserDefaults.h"
 #import "MSWrapperSdk.h"
-#import "MSUtil.h"
+#import "MSUtility+Application.h"
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/MobileCenter+Internal.h]: Sort import in alphabetical order.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Utility class that is used throughout the SDK.
+ * Application part.
+ */
+@interface MSUtility (Application)
+/**
+ * Get current application state.
+ *
+ * @discussion The application state may not be available anywhere. Application extensions doesn't have it for instance,
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.h]: Move `@discussion` below `@return`. You also need to add an empty line between different annotations.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Date.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Utility class that is used throughout the SDK.
+ * Date part.
+ */
+@interface MSUtility (Date)
+
+/**
+ * Return the current date (aka NOW) in ms.
+ *
+ * @discussion
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Date.h]: Here, too. The order would be `@param`, `@return`, `@warning` and then `@discussion`

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#import <Foundation/Foundation.h>
+#import "MSUtility.h"
+
+
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h]: Remove unnecessary empty line. One line should be enough.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ *  App environment
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h]: Can you remove one space before `App environment`?

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ */
+typedef NS_ENUM(NSInteger, MSEnvironment) {
+    /**
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h]: New lines before docs in this enum.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Utility class that is used throughout the SDK.
+ * Environment part.
+ */
+@interface MSUtility (Environment)
+
+/**
+ * Detect the environment that the app is running in.
+ * @return the MSEnvironment of the app.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.h]: New line.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.m [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#import "MSUtility+Environment.h"
+
+@implementation MSUtility (Environment)
+
+
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Environment.m]: Remove empty line.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+#define MS_USER_DEFAULTS [MSUserDefaults shared]
+#define MS_NOTIFICATION_CENTER [NSNotificationCenter defaultCenter]
+#define MS_DEVICE [UIDevice currentDevice]
+#define MS_UUID_STRING [[NSUUID UUID] UUIDString]
+#define MS_UUID_FROM_STRING(uuidString) [[NSUUID alloc] initWithUUIDString:uuidString]
+#define MS_LOCALE [NSLocale currentLocale]
+#define MS_CLASS_NAME_WITHOUT_PREFIX [NSStringFromClass([self class]) substringFromIndex:2]
+#define MS_IS_APP_EXTENSION [[[NSBundle mainBundle] executablePath] containsString:@".appex/"]
+
+
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility.h]: Remove empty line.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility.m [hunk 10]
```diff
@@ -0,0 +1,8 @@
+/*
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility.m]: We don't need a copyright. Remove.

File: MobileCenter/MobileCenterTests/MSUtilTests.m [hunk 11]
```diff
@@ -1,5 +1,5 @@
-#import "MSUtil.h"
-#import "MSUtilPrivate.h"
+#import "MSUtility+ApplicationPrivate.h"
+#import "MSUtility+Date.h"
 #import "OCMock.h"
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenterTests/MSUtilTests.m]: Can you change this to `#import<OCMock/OCMock.h>`?

File: Puppet/Puppet.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -792,6 +793,7 @@
 				INFOPLIST_FILE = Puppet/Info.plist;
 				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
+				OTHER_LDFLAGS = "-all_load";
```
Comment (TroubleMakerBen) [inline Puppet/Puppet.xcodeproj/project.pbxproj]: Correct me if I'm wrong, but this means that everyone who wants to include our SDK needs to add this flag.
I'm not okay with this.

Comment (bmourat) [inline Puppet/Puppet.xcodeproj/project.pbxproj]: Thanks for the article. 4th solution seems reasonable, so if everybody supports it, i will make necessary changes.

**Comment by bmourat** [comment]
Removed  need to add -all_load flag to target projects. Used 4th solution from this [article](https://medium.com/ios-os-x-development/categories-in-static-libraries-78e41f8ddb96#.15ww53e9w)

**Comment by lumaxis** [comment]
@bmourat Just FYI, we used a similar solution to #4 from that article in the Application Insights iOS SDK too, back in the day: https://github.com/Microsoft/ApplicationInsights-iOS/blob/develop/Classes/MSAICategoryContainer.h
🙂

**Comment by bmourat** [comment]
Thanks @lumaxis ! It's good to know :)

File: MobileCenter/MobileCenter/Internals/MobileCenter+Internal.h [hunk 13]
```diff
@@ -1,16 +1,17 @@
...
+#import "MSServiceAbstractInternal.h"
+#import "MSServiceInternal.h"
+#import "MSUtility+Application.h"
+#import "MSUtility+Date.h"
+#import "MSUtility+Environment.h"
 #import "MSWrapperSdk.h"
-#import "MSUtil.h"
 
-// Channel
+//Channel
```
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/MobileCenter+Internal.h]: You should restore the space here and add a `.` in the end `// Channel.` to be consistent with other comments.

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.h [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    MSApplicationStateUnknown
+};
+
+/**
+ * Utility class that is used throughout the SDK.
+ * Application part.
+ */
+@interface MSUtility (Application)
+/**
```
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.h]: Insert a new line before class documentation.

File: Puppet/Puppet.xcodeproj/project.pbxproj [hunk 15]
```diff
@@ -765,6 +765,7 @@
 				INFOPLIST_FILE = Puppet/Info.plist;
 				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
+				OTHER_LDFLAGS = "";
```
Comment (TroubleMakerBen) [inline Puppet/Puppet.xcodeproj/project.pbxproj]: Please remove those from the .xcproject file completely. Thx!

File: MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.m [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    }
+    return MSApplicationStateUnknown;
+}
+
++ (UIApplicationState)sharedAppState {
+    
+    // Compute selector at runtime for more discretion.
+    SEL sharedAppSel = NSSelectorFromString(@"sharedApplication");
+    return [[((UIApplication *(*)(id, SEL)) [[UIApplication class] methodForSelector:sharedAppSel])(
+                                                                                                    [UIApplication class], sharedAppSel) valueForKey:@"applicationState"] longValue];
```
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Utils/MSUtility+Application.m]: Does your code formater is well setup? This line looks like over the limit of 120 character defined in the `.clang-format` file at the root of the repo.

**Comment by TroubleMakerBen** [comment]
Looking good to me now.

**Pull request merged**